hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -2049,9 +2060,20 @@ def _pooled_kwargs(self, ir_type, kwargs):
 
     def _pop_ir_tree_node(self, ir_type: IRTypeName, kwargs: IRKWargsType) -> IRNode:
         assert self.ir_tree_nodes is not None
+
+        if self.ir_tree_nodes == []:
+            self.mark_overrun()
+
         node = self.ir_tree_nodes.pop(0)
-        assert node.ir_type == ir_type
-        assert kwargs == node.kwargs
+        if node.ir_type != ir_type or node.kwargs != kwargs:
+            # Unlike buffers, not every ir tree is a valid choice sequence. If
+            # we want to maintain determinism we don't have many options here
+            # beyond giving up when a modified tree becomes misaligned.
+            #
+            # For what it's worth, misaligned buffers — albeit valid — are rather
+            # unlikely to be *useful* buffers, so this isn't an enormous
+            # downgrade.
+            self.mark_overrun()
 
         return node
```

Comment:
Given that we have the same type-based assertions at every call site, I'd be inclined to put them inline here instead.

Reply:
referring to this? https://github.com/HypothesisWorks/hypothesis/blob/aa8a3a38df90a191109061ac72d631a79c9f0923/hypothesis-python/src/hypothesis/internal/conjecture/data.py#L1848-L1849
Unfortunately those are for mypy's benefit, or they would be inline here.

Reply:
Huh, fair enough.  I think it'd be possible to get around that with generics + overloads, but it'd be worse than status quo.